### PR TITLE
Some improvements

### PR DESCRIPTION
**Pull request by pbauer**
mainly handling BrokenModified and some UX-stuff.

### CODE REVIEW

**Comment by tschorr** [comment]
I'm still not happy with silently ignoring catalog inconsistencies.
The cmf_uid index is used to determine the working copy of a history. Incorrectly indexed content will show up as having no working copy in the histories list!
I'm quite sure that silently ignoring catalog inconsistencies will sooner or later lead to people deleting histories of **existing** working copies. That is definitely not what we want.
The correct way to deal with this is to rebuild/reindex the catalog before recalculating the statistics, even if it takes a long time.

**Comment by pbauer** [comment]
Ok, I'll remove the try-except around `getObject`. Are the other things ok? I have no idea how to create a `BrokenModified` in a test since it only happens when a class is removed (e.g. a revision of a document has a Assignment for a portlet that was removed later).

**Comment by tschorr** [comment]
Thanks. I'm happy to merge the rest, but we definitely need tests (goal is to further increase test coverage). Can't you just raise ``BrokenModified`` in your test code?
`
>>> from ZODB.broken import BrokenModified
>>> raise BrokenModified('foo')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
BrokenModified: foo
`

**Comment by tschorr** [comment]
I.e. use a mock that does the above?

File: src/collective/revisionmanager/browser/histories.pt [hunk 1]
```diff
@@ -110,6 +116,10 @@
                 <input tal:replace="structure context/@@authenticator/authenticator" />
 
             </form>
+
+        <script type="text/javascript"
```
Comment (hvelarde) [inline src/collective/revisionmanager/browser/histories.pt]: setting `type="text/javascript"` is not needed as is the default

File: src/collective/revisionmanager/browser/views.py [hunk 2]
```diff
@@ -166,10 +185,9 @@ def handle_save_action(self, action):
             return
         CheckAuthenticator(self.request)
         if not self.available():
-            self.status = _(
-                u'text_not_allowed_manage_server',
-                default=u'You are not allowed to manage the Zope server.'
-            )
+            msg = _(u'text_not_allowed_manage_server',
```
Comment (hvelarde) [inline src/collective/revisionmanager/browser/views.py]: follow coding conventions as before your changes

File: src/collective/revisionmanager/browser/views.py [hunk 3]
```diff
@@ -182,10 +200,9 @@ def handle_save_action(self, action):
     def handle_recalculate_stats(self, action):
         CheckAuthenticator(self.request)
         if not self.available():
-            self.status = _(
-                u'text_not_allowed_manage_server',
-                default=u'You are not allowed to manage the Zope server.'
-            )
+            msg = _(u'text_not_allowed_manage_server',
```
Comment (hvelarde) [inline src/collective/revisionmanager/browser/views.py]: follow coding conventions as before your changes

File: src/collective/revisionmanager/statscache.py [hunk 4]
```diff
@@ -56,6 +58,19 @@ def _save_retrieve(htool, hid, length):
                 except POSKeyError:
                     # bad luck - use previously defined fallback
                     pass
+
+        except BrokenModified as e:
+            log.warn('BrokenModified encountered trying to retrieve history {0}: {1}'.format(hid, e))  # noqa
```
Comment (hvelarde) [inline src/collective/revisionmanager/statscache.py]: use `# noqa: E501` instead

File: src/collective/revisionmanager/statscache.py [hunk 5]
```diff
@@ -44,8 +46,8 @@ def _save_retrieve(htool, hid, length):
         """
         try:
             wrapper = htool.retrieve(hid).object
-        except POSKeyError:
-            log.warn('POSKeyError encountered trying to retrieve history {}'.format(hid))  # noqa
+        except POSKeyError as e:
+            log.warn('POSKeyError encountered trying to retrieve history {0}: {1}'.format(hid, e))  # noqa
```
Comment (hvelarde) [inline src/collective/revisionmanager/statscache.py]: use `# noqa: E501` instead

File: src/collective/revisionmanager/tests/test_revmgmt.py [hunk 6]
```diff
@@ -319,7 +335,7 @@ def test_retrieve_histories_with_poskeyerror_no_blob_file(self):
             cache.refresh()
             log.check(
                 ('collective.revisionmanager.statscache', 'WARNING',
-                 'POSKeyError encountered trying to retrieve history 1'),
+                 "POSKeyError encountered trying to retrieve history 1: 'No blob file'"),  # noqa
```
Comment (hvelarde) [inline src/collective/revisionmanager/tests/test_revmgmt.py]: quotes should be the other way around and use `# noqa: E501` instead

File: src/collective/revisionmanager/tests/test_revmgmt.py [hunk 7]
```diff
@@ -376,11 +392,58 @@ def test_working_copy_deleted_and_poskeyerror(self):
...
+        # the time needed to calculate stats may vary
+        self.assertTrue(float(got['summaries']['time']) < 1)
+        self.assertEqual(len(expected['histories']), len(got['histories']))
+        for idx in range(len(expected['histories'])):
+            e = expected['histories'][idx]
+            g = got['histories'][idx]
+            for k, v in e.items():
+                self.assertEqual(g[k], v)
+            # The actual size is not important and we want robust tests,
+            self.assertTrue(g['size'] > 0)
```
Comment (hvelarde) [inline src/collective/revisionmanager/tests/test_revmgmt.py]: use `self.assertGreater()` instead

File: src/collective/revisionmanager/tests/test_revmgmt.py [hunk 8]
```diff
@@ -376,11 +392,58 @@ def test_working_copy_deleted_and_poskeyerror(self):
...
+                'wcinfos': [{
+                    'url': None,
+                    'path': 'no working copy, object id: obj',
+                    'portal_type': 'Dummy'}],
+                'size_state': 'approximate'}]}
+        got = cache
+        for k, v in expected['summaries'].items():
+            self.assertEqual(got['summaries'][k], v)
+        # the time needed to calculate stats may vary
+        self.assertTrue(float(got['summaries']['time']) < 1)
```
Comment (hvelarde) [inline src/collective/revisionmanager/tests/test_revmgmt.py]: use self.assertLess() instead

File: src/collective/revisionmanager/tests/test_revmgmt.py [hunk 9]
```diff
@@ -376,11 +392,58 @@ def test_working_copy_deleted_and_poskeyerror(self):
...
+        that was removed later) BrokenModified is Raised.
+        """
+        # decorate portal_historiesstorage retrieve method with mock
+        self.portal_storage.retrieve = mock_retrieve_with_brokenmodified((1,))(self.portal_storage.retrieve)  # noqa
+        cache = getUtility(IHistoryStatsCache)
+        with LogCapture(level=WARN) as log:
+            cache.refresh()
+            log.check(
+                ('collective.revisionmanager.statscache', 'WARNING',
+                 "BrokenModified encountered trying to retrieve history 1: <persistent broken some.package.portlet.Assignment instance '\x00\x00\x00\x00\x00H\xf3\xc2'"),  # noqa
```
Comment (hvelarde) [inline src/collective/revisionmanager/tests/test_revmgmt.py]: use `# noqa: E501` instead

File: src/collective/revisionmanager/tests/test_revmgmt.py [hunk 10]
```diff
@@ -376,11 +392,58 @@ def test_working_copy_deleted_and_poskeyerror(self):
...
         got = [h for h in cache['histories'] if h['history_id'] == 2][0]['wcinfos']  # noqa
         self.assertEqual(expected, got)
+
+    def test_retrieve_histories_with_brokenmodified(self):
+        """When a class is removed that a revision relies upon
+        (e.g. when a revision of a document has a Assignment for a portlet
+        that was removed later) BrokenModified is Raised.
+        """
+        # decorate portal_historiesstorage retrieve method with mock
+        self.portal_storage.retrieve = mock_retrieve_with_brokenmodified((1,))(self.portal_storage.retrieve)  # noqa
```
Comment (hvelarde) [inline src/collective/revisionmanager/tests/test_revmgmt.py]: use `# noqa: E501` instead

File: src/collective/revisionmanager/tests/test_revmgmt.py [hunk 11]
```diff
@@ -376,11 +392,58 @@ def test_working_copy_deleted_and_poskeyerror(self):
             cache.refresh()
             log.check(
                 ('collective.revisionmanager.statscache', 'WARNING',
-                 'POSKeyError encountered trying to retrieve history 2'),
+                 "POSKeyError encountered trying to retrieve history 2: 'No blob file'"),  # noqa
```
Comment (hvelarde) [inline src/collective/revisionmanager/tests/test_revmgmt.py]: use `# noqa: E501` instead

File: src/collective/revisionmanager/tests/test_revmgmt.py [hunk 12]
```diff
@@ -276,6 +278,20 @@ def wrapped_m(*passedargs, **kwds):
     return decorate
 
 
+def mock_retrieve_with_brokenmodified(args):
+    """ Very specific decorator to make ZVCStorageTool.retrieve
```
Comment (hvelarde) [inline src/collective/revisionmanager/tests/test_revmgmt.py]: docstrings should not start with spaces

**Comment by pbauer** [comment]
@tschorr done

**Comment by tschorr** [comment]
@pbauer there is still a decrease in test coverage :-(
More specifically, could you add some tests for the new code in views.py?

**Comment by pbauer** [comment]
done. not perfect but better that before.

**Comment by tschorr** [comment]
@pbauer there is still a decrease in test coverage :-(
More specifically, could you add some tests for the new code in views.py?

**Comment by tschorr** [comment]
9.3 % increase in test coverage, that's a huge improvement. Thanks!

**Comment by hvelarde** [comment]
@tschorr using squash instead of merge is better sometimes.

**Comment by pbauer** [comment]
@hvelarde done

**Pull request merged**